### DISCUSSION TITLE
Forwarding props with store and context

### DISCUSSION THREAD

**Discussion by johsunds** [pmndrs/zustand#1390] [Q&A]
Hi!

TL;DR: I want to update a zustand store when the props passed to my component changes. What would be the best way to do this? `UNSAFE_componentWillReceiveProps`? `useEffect`?

I have a rather complex component in which I'm using a zustand store to manage the state using the context pattern described [here](https://github.com/pmndrs/zustand#react-context). In addition to managing the components state, I also want to use the store to forward props, so that I can use the store selector to grab only the props I need further down the tree:

(not my exact code, but hopefully you get the point)

```
function MyComponent(props: { myProp: any }) {
  const store = useConstant(() => createStore(props))
  return (
    <StoreProvider value={store}>
      // MyComponentImpl is wrapped in React.memo
      <MyComponentImpl></MyComponentImpl>
    </StoreProvider>
  );
}

// Somewhere deep inside MyComponentImpl
function MyDeeplyNestedSubComponent() {
  const store = React.useContext(StoreContext)
  const myProp = useStore(store, (state) => state.myProp)
}
```

This all works fine, but then there's the question of how to update the prop in the store when the prop passed to `MyComponent` changes. Doing the update in `useEffect` is the obvious choice:

```
function MyComponent(props: { myProp: any }) {
  const store = useConstant(() => createStore(props));

  React.useEffect(() => {
    store.setState(props);
  }, [props]);
  
  return (
    <StoreProvider value={store}>
      <MyComponentImpl></MyComponentImpl>
    </StoreProvider>
  );
}
```

but effects run later in the component lifecycle right? so it would not emulate passing the props normally? An alternative I tried is to rewrite it as a class component and do the `store.setState(props)` in `UNSAFE_componentWillReceiveProps`. This seems to work, but frankly the `UNSAFE` prefix is scaring me a bit. 😁 

My questions:

- Is this a horrible idea? Props aren't really state in the sense that they shouldn't be mutated from inside the component, so it does feel a bit iffy putting it in the store. However, I have a lot of props that need to be passed down many levels, so I would like to avoid prop drilling. I could just pass them down in another context without the store, but then I lose the selector functionality.
- If it is not a horrible idea: how should I update the store on prop changes? `useEffect`? `UNSAFE_componentWillReceiveProps`? anywhere else?

Thanks!

**Comment by dai-shi** [comment]
This is an unfortunate mismatch between react state/props and global state.

If you want to use zustand store, useEffect would be a only recommended solution.
But, to avoid prop drilling, react context should be the right tool.
I also develop a library if you are interested: [use-context-selector](https://github.com/dai-shi/use-context-selector)

**Reply by johsunds** [reply]
Right now I have a (seemingly) working solution with `UNSAFE_componentWillReceiveProps` (where I update the store state in that callback when the props have changed), are there any obvious issues with that in your opinion?

[use-context-selector](https://github.com/dai-shi/use-context-selector) looks to be exactly what I'm after, since I'm primarily using `zustand` for the selector at this point, so I might just switch over to that. Thanks for your amazing contributions!
